sympy/functions/elementary/trigonometric.py

```diff
@@ -1527,10 +1527,18 @@ def eval(cls, arg):
                     elif cls._is_even:
                         return -cls(narg)
 
-        t = cls._reciprocal_of.eval(arg)
         if hasattr(arg, 'inverse') and arg.inverse() == cls:
             return arg.args[0]
-        return 1/t if t != None else t
+
+        t = cls._reciprocal_of.eval(arg)
+        if t == None:
+            return t
+        elif isinstance(t, cls._reciprocal_of):
+            return (1/t).rewrite(cls)
```

Comment:
what is an example of a (1/t) that gets rewritten (here and below)? I'm just wondering if the rewrite is necessary.

Reply:
The statements help write `1/cos(arg)` as `sec(arg)` (and `1/sin(arg)` as `csc(arg)`). 
The current behaviour doesn't account for this (ie when arguments are specified for `sec` and `csc`), that's why I think rewrite is required. 
As in the given example,
`sec(2*pi + 11)` gets reduced to `1/cos(11)` (currently) and this PR reduces it to `sec(11)`. 
Similarly,
`sec(pi + 11)` gets reduced to `-1/cos(11)` (currently) and this PR reduces it to `-sec(11)`.

Reply:
I do not understand why sec.eval should ever return a result in terms of cos.

Reply:
If I understand correctly, once the variable `t` is evaluated, directly returning `1/t` causes this issue, where `t` can be an instance of `cos` (or `sin`), which doesn't automatically reduce to `sec` (or `csc`).

Reply:
I guess the idea is that something like `sec(pi/2)` can use `cos.eval` to compute the value. But `cos` also does automatic simplifications where it returns a simpler cos (like `cos(2*pi + x)`). It also actually does some simplifications where it returns a different trig function, like `cos(pi/2 - x)`. `sec(pi/2 - x)` should arguably return `csc(x)`. 

The trick used here feels convoluted, and could be susceptible to infinite recursion. It's definitely susceptible to the specific types of simplifications done by cos (it handles cos returning cos and cos returning -cos, but not cos returning sin). But this is also probably the simplest fix that doesn't require refactoring cos.eval so that sec.eval could use it more directly. 

If refactoring were to be done (I'm not saying it needs to be done in this pull request), my suggestion would be to factor out the different kinds of automatic simplifications into separate sub methods. For instance, simplifications that result in a value would go in one method. Simplifications that result in a function of the same type (`cos(x + 2*pi)`) would go in another, and simplifications that result in a function of a different type would go in yet another (`cos(pi/2 - x)`). That way `sec.eval` could use these directly (reciprocate for the first, use directly for the second, use replacing func with func._reciprocal_of in the third, and so on). The trig eval functions are also currently somewhat littered with SetExpr and AccumBounds stuff, so that could be refactored as well (probably we need `_eval_sin` type dispatching).

Reply:
> `sec(pi/2 - x)` should arguably return `csc(x)`
> it handles cos returning cos and cos returning -cos, but not cos returning sin

Yes, indeed a good catch.

> my suggestion would be to factor out the different kinds of automatic simplifications into separate submethods

Can't we simply identify the instance of the variable as `+/- sin/cos` and deal directly with their `func._reciprocal_of`? What would be the advantage of having submethods?

Reply:
It would clean up the code.

Reply:
Okay, so should this PR be modified to account for `+/- sin` (for `sec`), if refactoring is to be a part of different PR?

Reply:
Refactoring in a separate PR seems better. It would be a much larger task.

Reply:
Okay, so I will make the desired changes for this PR.
@asmeurer @smichr Thank you for the review.